### DISCUSSION TITLE
How to call a new defined cell/function in GDSfactory

### DISCUSSION THREAD

**Discussion by GiaQuyetNgo** [gdsfactory/gdsfactory#2926] [Q&A] [answered]
Hi everyone,
I want to create a complex photonic structure including several couplers, add-drop microrings, bends, and straight lines with various parameters such as gaps, width, length, and radius. I am trying to split the final structure into smaller components and connect them in the final step.
However, when I created the new Python file in /gdsfactory/components/ with defined cells using the cell decorator, I could not call them in another Python file to make the final structure. It gave the error: AttributeError: module 'gdsfactory.components' has no attribute 'directional_coupler', as an example. The existing cells from GDS factory by default worked well, but it won't work if I change the cell name and/or file name.
The reason I don't want to make the entire thing in a single Python file is to avoid the complexity of the connection of so many small elements.
Thank you very much for your help.
Kind regards
Quyet

**Comment by sebastian-goeldi** [comment, marked as answer]
Hi Quyet,

Placing new components in gdsfactory/components only makes sense if you wish to make them available for all gdsfactory users, which I assume you don't want to do.

It's best if you make your own python package (or at least collection of files).

For how to do that, you can take one of the open source gdsfactory packages as a template, e.g. the [cornerstone](https://github.com/gdsfactory/cspdk/) should be relatively up to date.

As for how python packages work, I would suggest to read one of the guides for doing so, e.g. the official one https://packaging.python.org/en/latest/tutorials/packaging-projects/

If you think that's too complicated (a package is what ultimately allows you to do `pip install gdsfactory`, so there's some configuration on top, namely `pyproject.toml`), you can do it by putting all your files into one directory and then importing those, note: this will only work as long as you use the folder with the python files as your current working directory:

<details open>
<summary>file1.py</summary> 

```python
import gdsfactory as gf

@gf.cell
def my_mmi() -> gf.Component:
    c = gf.Component()
    # ...
    return c
```

</details>

<details open>
<summary>file2.py</summary>

```python
from file1 import my_mmi

@gf.cell
def my_mzi() -> gf.Component:
    c = gf.Component()
    mmi1 = c << my_mmi()
    mmi2 = c << my_mmi()
    # ...
    return c
```

</details

**Reply by GiaQuyetNgo** [reply]
Hi Sebastian. Thank you for your helpful support and great examples.
